drivers/vulkan/rendering_device_driver_vulkan.cpp

```diff
@@ -5268,6 +5562,374 @@ RDD::PipelineID RenderingDeviceDriverVulkan::render_pipeline_create(
+	accel_info->build_info.pGeometries = &accel_info->geometry;
+	accel_info->build_info.geometryCount = 1;
+	accel_info->build_info.flags = VK_BUILD_ACCELERATION_STRUCTURE_PREFER_FAST_TRACE_BIT_KHR;
+
+	VkAccelerationStructureBuildSizesInfoKHR size_info = {};
+	size_info.sType = VK_STRUCTURE_TYPE_ACCELERATION_STRUCTURE_BUILD_SIZES_INFO_KHR;
+
+	vkGetAccelerationStructureBuildSizesKHR(vk_device, VK_ACCELERATION_STRUCTURE_BUILD_TYPE_DEVICE_KHR, &accel_info->build_info, &max_primitive_count, &size_info);
+	_acceleration_structure_create(VK_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL_KHR, size_info, accel_info);
+
+	return AccelerationStructureID(accel_info);
+#else
+	return AccelerationStructureID();
+#endif
+}
+
+#if !(defined(MACOS_ENABLED) || defined(IOS_ENABLED))
+static _FORCE_INLINE_ void _store_transform_transposed_3x4(const Transform3D &p_mtx, VkTransformMatrixKHR &r_mtx) {
+	r_mtx.matrix[0][0] = p_mtx.basis.rows[0][0];
+	r_mtx.matrix[0][1] = p_mtx.basis.rows[0][1];
+	r_mtx.matrix[0][2] = p_mtx.basis.rows[0][2];
+	r_mtx.matrix[0][3] = p_mtx.origin.x;
+	r_mtx.matrix[1][0] = p_mtx.basis.rows[1][0];
+	r_mtx.matrix[1][1] = p_mtx.basis.rows[1][1];
+	r_mtx.matrix[1][2] = p_mtx.basis.rows[1][2];
+	r_mtx.matrix[1][3] = p_mtx.origin.y;
+	r_mtx.matrix[2][0] = p_mtx.basis.rows[2][0];
+	r_mtx.matrix[2][1] = p_mtx.basis.rows[2][1];
+	r_mtx.matrix[2][2] = p_mtx.basis.rows[2][2];
+	r_mtx.matrix[2][3] = p_mtx.origin.z;
+}
+#endif
+
+RDD::AccelerationStructureID RenderingDeviceDriverVulkan::tlas_create(const LocalVector<AccelerationStructureID> &p_blases, const Vector<Transform3D> &p_transforms) {
+#if !(defined(MACOS_ENABLED) || defined(IOS_ENABLED))
+	ERR_FAIL_COND_V_MSG(p_blases.size() != p_transforms.size(), AccelerationStructureID(), "Blases and transforms vectors must have the same size.");
+
+	AccelerationStructureInfo *accel_info = VersatileResource::allocate<AccelerationStructureInfo>(resources_allocator);
+
+	for (uint32_t i = 0; i < p_blases.size(); ++i) {
+		const AccelerationStructureID &blas = p_blases[i];
+		AccelerationStructureInfo *blas_info = (AccelerationStructureInfo *)blas.id;
+
+		VkAccelerationStructureInstanceKHR instance = {};
+		_store_transform_transposed_3x4(p_transforms[i], instance.transform);
+		instance.instanceCustomIndex = i;
+		instance.mask = 0xFF;
+		instance.accelerationStructureReference = buffer_get_device_address(blas_info->buffer);
+		instance.instanceShaderBindingTableRecordOffset = 0;
+		instance.flags = VK_GEOMETRY_INSTANCE_TRIANGLE_FACING_CULL_DISABLE_BIT_KHR;
+
+		accel_info->instances.push_back(instance);
+	}
+
+	uint32_t instance_count = accel_info->instances.size();
+	VkDeviceAddress instances_buffer_address = 0;
+
+	if (instance_count > 0) {
+		uint32_t instances_size = instance_count * sizeof(accel_info->instances[0]);
+		accel_info->instances_buffer = buffer_create(instances_size, BUFFER_USAGE_DEVICE_ADDRESS_BIT | BUFFER_USAGE_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT, MEMORY_ALLOCATION_TYPE_CPU);
```

Comment:
We should go one step further here and make the transforms buffer itself an input argument. As it stands, the only way to modify the TLAS will be through the CPU, a limitation that is not part of the APIs but one imposed by the RDD instead.

This closes the door to potential improvements such as computing the transform from another source in the GPU, which would be the case with something like GPU particles.